ReconcilerUtils.setSpec() cannot find the spec if it is a Custom Resource

## Bug Report

#### What did you do?

<!-- A clear and concise description of the steps you took (or insert a code snippet). -->
When updating an instance of a `Custom Resource` we get an exception in the [ReconcilerUtils.setSpec(HasMetadata resource, Object spec)](https://github.com/java-operator-sdk/java-operator-sdk/blob/fb9dc1aeef2a5b0c9faf69286d39f16409f332f7/operator-framework-core/src/main/java/io/javaoperatorsdk/operator/ReconcilerUtils.java#L104):
```java
io.javaoperatorsdk.operator.AggregatedOperatorException: Exception during workflow.
	at io.javaoperatorsdk.operator.processing.dependent.workflow.WorkflowReconcileResult.createFinalException(WorkflowReconcileResult.java:54) ~[operator-framework-core-3.2.0.jar:na]
	at io.javaoperatorsdk.operator.processing.dependent.workflow.WorkflowReconcileResult.throwAggregateExceptionIfErrorsPresent(WorkflowReconcileResult.java:48) ~[operator-framework-core-3.2.0.jar:na]
	at io.javaoperatorsdk.operator.processing.dependent.workflow.Workflow.reconcile(Workflow.java:56) ~[operator-framework-core-3.2.0.jar:na]
	at no.fintlabs.FlaisReconiler.reconcile(FlaisReconiler.java:50) ~[flais-operator-starter-1.0.0-rc-4.jar:na]
	at no.fintlabs.FlaisReconiler.reconcile(FlaisReconiler.java:20) ~[flais-operator-starter-1.0.0-rc-4.jar:na]
	at io.javaoperatorsdk.operator.processing.Controller$1.execute(Controller.java:130) ~[operator-framework-core-3.2.0.jar:na]
	at io.javaoperatorsdk.operator.processing.Controller$1.execute(Controller.java:88) ~[operator-framework-core-3.2.0.jar:na]
	at io.javaoperatorsdk.operator.api.monitoring.Metrics.timeControllerExecution(Metrics.java:197) ~[operator-framework-core-3.2.0.jar:na]
	at io.javaoperatorsdk.operator.processing.Controller.reconcile(Controller.java:87) ~[operator-framework-core-3.2.0.jar:na]
	at io.javaoperatorsdk.operator.processing.event.ReconciliationDispatcher.reconcileExecution(ReconciliationDispatcher.java:130) ~[operator-framework-core-3.2.0.jar:na]
	at io.javaoperatorsdk.operator.processing.event.ReconciliationDispatcher.handleReconcile(ReconciliationDispatcher.java:110) ~[operator-framework-core-3.2.0.jar:na]
	at io.javaoperatorsdk.operator.processing.event.ReconciliationDispatcher.handleDispatch(ReconciliationDispatcher.java:81) ~[operator-framework-core-3.2.0.jar:na]
	at io.javaoperatorsdk.operator.processing.event.ReconciliationDispatcher.handleExecution(ReconciliationDispatcher.java:54) ~[operator-framework-core-3.2.0.jar:na]
	at io.javaoperatorsdk.operator.processing.event.EventProcessor$ReconcilerExecutor.run(EventProcessor.java:395) ~[operator-framework-core-3.2.0.jar:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[na:na]
	at java.base/java.lang.Thread.run(Thread.java:833) ~[na:na]
Caused by: java.lang.IllegalStateException: No spec found on resource
	at io.javaoperatorsdk.operator.ReconcilerUtils.setSpec(ReconcilerUtils.java:112) ~[operator-framework-core-3.2.0.jar:na]
	at io.javaoperatorsdk.operator.processing.dependent.kubernetes.GenericResourceUpdatePreProcessor$3.updateClonedActual(GenericResourceUpdatePreProcessor.java:40) ~[operator-framework-core-3.2.0.jar:na]
	at io.javaoperatorsdk.operator.processing.dependent.kubernetes.GenericResourceUpdatePreProcessor.replaceSpecOnActual(GenericResourceUpdatePreProcessor.java:48) ~[operator-framework-core-3.2.0.jar:na]
	at io.javaoperatorsdk.operator.processing.dependent.kubernetes.KubernetesDependentResource.update(KubernetesDependentResource.java:132) ~[operator-framework-core-3.2.0.jar:na]
	at io.javaoperatorsdk.operator.processing.dependent.kubernetes.CRUDKubernetesDependentResource.update(CRUDKubernetesDependentResource.java:17) ~[operator-framework-core-3.2.0.jar:na]
	at io.javaoperatorsdk.operator.processing.dependent.AbstractDependentResource.handleUpdate(AbstractDependentResource.java:111) ~[operator-framework-core-3.2.0.jar:na]
	at io.javaoperatorsdk.operator.processing.dependent.kubernetes.KubernetesDependentResource.handleUpdate(KubernetesDependentResource.java:119) ~[operator-framework-core-3.2.0.jar:na]
	at io.javaoperatorsdk.operator.processing.dependent.kubernetes.KubernetesDependentResource.handleUpdate(KubernetesDependentResource.java:32) ~[operator-framework-core-3.2.0.jar:na]
	at io.javaoperatorsdk.operator.processing.dependent.AbstractDependentResource.reconcile(AbstractDependentResource.java:50) ~[operator-framework-core-3.2.0.jar:na]
	at io.javaoperatorsdk.operator.processing.dependent.workflow.WorkflowReconcileExecutor$NodeReconcileExecutor.run(WorkflowReconcileExecutor.java:166) ~[operator-framework-core-3.2.0.jar:na]
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539) ~[na:na]
	at java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:264) ~[na:na]
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java) ~[na:na]
	... 3 common frames omitted
Caused by: java.lang.NoSuchMethodException: no.fintlabs.onepassword.OnePasswordCrd.setSpec(no.fintlabs.onepassword.OnePasswordSpec)
	at java.base/java.lang.Class.getMethod(Class.java:2227) ~[na:na]
	at io.javaoperatorsdk.operator.ReconcilerUtils.setSpec(ReconcilerUtils.java:109) ~[operator-framework-core-3.2.0.jar:na]
	... 15 common frames omitted
```

This is only happening when we update, not on create.

This is the `desired` method from the dependent resource class. This is how I create the CRD instance.
```java
    @Override
    protected OnePasswordCrd desired(FlaisApplicationCrd primary, Context<FlaisApplicationCrd> context) {

        OnePasswordCrd onePasswordCrd = new OnePasswordCrd();
        onePasswordCrd.getMetadata().setLabels(LabelFactory.updateRecommendedLabels(primary));
        onePasswordCrd.getMetadata().setName(primary.getMetadata().getName());
        onePasswordCrd.getMetadata().setNamespace(primary.getMetadata().getNamespace());
        onePasswordCrd.getSpec().setItemPath(primary.getSpec().getOnePassword().getItemPath());

        return onePasswordCrd;
    }
```

This is the CRD class.
```java
@Group("onepassword.com")
@Version("v1")
@Kind("OnePasswordItem")
public class OnePasswordCrd extends CustomResource<OnePasswordSpec, Void> implements Namespaced {
    @Override
    protected OnePasswordSpec initSpec() {
        return new OnePasswordSpec();
    }
}
```

This is the spec class.
```java
@Data
public class OnePasswordSpec {
    private String itemPath;
}
```

One can see the source code [here](https://github.com/FINTLabs/flaiserator/tree/FLAIS-9-stotte-for-one-password/src/main/java/no/fintlabs)

#### What did you expect to see?

No exception and the resource updated.

#### What did you see instead? Under which circumstances?

See the description further up.
#### Environment

**Kubernetes cluster type:**

Azure Kubernetes Service

`io.javaoperatorsdk.operator-framework:3.2.0`


`openjdk 17.0.2 2022-01-18 LTS`


```
Client Version: v1.24.0
Kustomize Version: v4.5.4
Server Version: v1.22.11
```

#### Possible Solution

What happens is that if it is a CustomResource we're not able to get the setSpec() method. 
The reason seems to be related to reflection. If I do resource.getClass().getMethod("setSpec", Object.class) 
instead of resource.getClass().getMethod("setSpec", spec.getClass()) it seems to work.

It might be a solution to check if the `resource` extends `CustomResource` and then do
`resource.getClass().getMethod("setSpec", Object.class)`

#### Additional context

<!-- Add any other context about the problem here. -->

--- a/operator-framework-core/src/main/java/io/javaoperatorsdk/operator/ReconcilerUtils.java
+++ b/operator-framework-core/src/main/java/io/javaoperatorsdk/operator/ReconcilerUtils.java
@@ -12,6 +12,7 @@ import java.util.stream.Collectors;
 
 import io.fabric8.kubernetes.api.builder.Builder;
 import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.fabric8.kubernetes.client.CustomResource;
 import io.fabric8.kubernetes.client.KubernetesClientException;
 import io.fabric8.kubernetes.client.utils.Serialization;
 import io.javaoperatorsdk.operator.api.reconciler.Constants;
@@ -103,7 +104,9 @@ public class ReconcilerUtils {
 
   public static Object setSpec(HasMetadata resource, Object spec) {
     try {
-      Method setSpecMethod = resource.getClass().getMethod("setSpec", spec.getClass());
+      Class<? extends HasMetadata> resourceClass = resource.getClass();
+      Method setSpecMethod =
+          resource.getClass().getMethod("setSpec", getSpecClass(resourceClass, spec));
       return setSpecMethod.invoke(resource, spec);
     } catch (NoSuchMethodException | InvocationTargetException | IllegalAccessException e) {
       throw new IllegalStateException("No spec found on resource", e);
@@ -165,6 +168,15 @@ public class ReconcilerUtils {
       }
     }
     return false;
+  }
+
+  // CustomResouce has a parameterized parameter type
+  private static Class getSpecClass(Class<? extends HasMetadata> resourceClass, Object spec) {
+    if (CustomResource.class.isAssignableFrom(resourceClass)) {
+      return Object.class;
+    } else {
+      return spec.getClass();
+    }
   }
 
 }

--- a/operator-framework-core/src/test/java/io/javaoperatorsdk/operator/ReconcilerUtilsTest.java
+++ b/operator-framework-core/src/test/java/io/javaoperatorsdk/operator/ReconcilerUtilsTest.java
@@ -99,6 +99,19 @@ class ReconcilerUtilsTest {
   }
 
   @Test
+  void setsSpecCustomResourceWithReflection() {
+    Tomcat tomcat = new Tomcat();
+    tomcat.setSpec(new TomcatSpec());
+    tomcat.getSpec().setReplicas(5);
+    TomcatSpec newSpec = new TomcatSpec();
+    newSpec.setReplicas(1);
+
+    ReconcilerUtils.setSpec(tomcat, newSpec);
+
+    assertThat(tomcat.getSpec().getReplicas()).isEqualTo(1);
+  }
+
+  @Test
   void loadYamlAsBuilder() {
     DeploymentBuilder builder =
         ReconcilerUtils.loadYaml(DeploymentBuilder.class, getClass(), "deployment.yaml");
@@ -134,7 +147,19 @@ class ReconcilerUtilsTest {
   @Group("tomcatoperator.io")
   @Version("v1")
   @ShortNames("tc")
-  private static class Tomcat extends CustomResource<Void, Void> implements Namespaced {
+  private static class Tomcat extends CustomResource<TomcatSpec, Void> implements Namespaced {
 
+  }
+
+  private class TomcatSpec {
+    private Integer replicas;
+
+    public Integer getReplicas() {
+      return replicas;
+    }
+
+    public void setReplicas(Integer replicas) {
+      this.replicas = replicas;
+    }
   }
 }

